### ISSUE TITLE
spurious duplicate entry in completion menu when 'lazyredraw' is set

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#10601]
**Steps to reproduce**

Open an xterm terminal, then run this shell command:

    vim -Nu NONE -S <(tee <<'EOF'
        vim9script
        &lazyredraw = true
        var completions = ['xaaa', 'xbbb', 'xccc']
        $'x {'y'->repeat(&columns - len('xaaa'))}'->setline(1)
        ['']->repeat(completions->len() + 1)->append('.')
        inoremap <F3> <ScriptCmd>completions->complete(1)<CR>
        feedkeys("a\<F3>")
    EOF
    )

The completion `xccc` is displayed twice at the bottom of the pum.

**Expected behavior**

The completion `xccc` is displayed only once at the bottom of the pum, because we've passed the `'xccc'` string only once to the `complete()` function.

**Version of Vim**

8.2 Included patches: 1-5141

**Environment**

Operating system: Ubuntu 20.04.4 LTS
Terminal: xterm
Value of $TERM: xterm-256color
Shell: GNU bash, version 5.0.17

**Screenshot**

![gif](https://user-images.githubusercontent.com/8505073/174686621-4aa16b69-7ca5-4017-b6d9-73ef4e847d36.gif)

**Additional context**

Regression introduced in patch [8.2.0202](https://github.com/vim/vim/releases/tag/v8.2.0202).

---

The issue disappears if `'lazyredraw'` is reset.

---

Pressing `<C-N>` repeatedly to cycle beyond the last completion makes the spurious `xccc` disappear (I guess that's because it causes a redraw of the pum).

---

I can reproduce in st.
I cannot reproduce in gnome-terminal.
I can – intermittently – reproduce in kitty, but I think it requires `'t_te'` and `'t_ti'` to be configured so that the window icon and title are saved/restored:

    &t_te = "\<Esc>[23;0;0t"
    &t_ti = "\<Esc>[22;0;0t"

**Comment by lacygoill** [comment]
> I cannot reproduce it.

I can consistently reproduce in xterm, with no shell config (`$ bash --norc --noprofile`), no terminal config (empty `~/.Xresources`), and starting the terminal from the whisker menu (aka the XFCE desktop application launcher).

---

  > It might be that a redraw happens on startup when the xterm version string is received.

I encountered this issue while Vim had already been running for some time; not on startup.  However, I've checked again, and it seems the current example is only able to reproduce on startup.  Maybe I'll try to find another example where it reproduces even after the startup has completed.  That's going to take quite some time; the current example was already hard enough to find.

   > Generally setting 'lazyredraw' should not be set when doing completion, it may skip some redrawing.

I have no idea how I would temporarily reset `'lazyredraw'` (a global option), and only while a completion is active.  Maybe with an autocmd, but that looks like a hack.

**Comment by lacygoill** [comment]
> Maybe I'll try to find another example where it reproduces even after the startup has completed.

Found it: We need to append the `<C-N>` keycode at the end of the rhs of the mapping, and get rid of the `feedkeys()`.

Write this in `/tmp/t.vim`:
```vim
vim9script
&lazyredraw = true
var completions = ['xaaa', 'xbbb', 'xccc']
$'x {'y'->repeat(&columns - len('xaaa'))}'->setline(1)
['']->repeat(completions->len() + 1)->append('.')
inoremap <F3> <ScriptCmd>completions->complete(1)<CR><C-N>
```
Start an xterm terminal, and Vim without config:

    $ vim -Nu NONE

Source the script (after startup):

    :source /tmp/t.vim

Finally, press:

   - `a` to enter insert mode right after the `x` character
   - `<F3>` to complete `x`

**Comment by lacygoill** [comment]
With the new example, I bisected the issue to the patch [8.1.0768](https://github.com/vim/vim/releases/tag/v8.1.0768).

**Comment by brammool** [comment]
>    > I cannot reproduce it.
> 
> I can consistently reproduce in xterm, with no shell config (`$ bash --norc --noprofile`), no terminal config (empty `~/.Xresources`), and starting the terminal from the whisker menu (aka the XFCE desktop application launcher).
> 
> ---
> 
>   > It might be that a redraw happens on startup when the xterm version string is received.
> 
> I encountered this issue while Vim had already been running for some time; not on startup.  However, I've checked again, and it seems the current example is only able to reproduce on startup.  Maybe I'll try to find another example where it reproduces even after the startup has completed.  That's going to take quite some time; the current example was already hard enough to find.
> 
>    > Generally setting 'lazyredraw' should not be set when doing completion, it may skip some redrawing.
> 
> I have no idea how I would temporarily reset `'lazyredraw'` (a global
> option), and only while a completion is active.  Maybe with an
> autocmd, but that looks like a hack.

Normally 'lazyredraw' should not be set, it is not normally needed.
If there are some operations where you don't want the screen to be
updated, when some things flash by, then set it with that operation, but
reset it afterwards.


-- 
Scientists decoded the first message from an alien civilization:
        SIMPLY SEND 6 TIMES 10 TO THE 50 ATOMS OF HYDROGEN TO THE STAR
SYSTEM AT THE TOP OF THE LIST, CROSS OFF THAT STAR SYSTEM, THEN PUT
YOUR STAR SYSTEM AT THE BOTTOM OF THE LIST AND SEND IT TO 100 OTHER
STAR SYSTEMS.  WITHIN ONE TENTH GALACTIC ROTATION YOU WILL RECEIVE
ENOUGH HYDROGREN TO POWER YOUR CIVILIZATION UNTIL ENTROPY REACHES ITS
MAXIMUM!  IT REALLY WORKS!

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
> > Maybe I'll try to find another example where it reproduces even
> > after the startup has completed.
> 
> Found it: We need to append the `<C-N>` keycode at the end of the rhs
> of the mapping, and get rid of the `feedkeys()`.
> 
> Write this in `/tmp/t.vim`:
> ```vim
> vim9script
> &lazyredraw = true
> var completions = ['xaaa', 'xbbb', 'xccc']
> $'x {'y'->repeat(&columns - len('xaaa'))}'->setline(1)
> ['']->repeat(completions->len() + 1)->append('.')
> inoremap <F3> <ScriptCmd>completions->complete(1)<CR><C-N>
> ```
> Start an xterm terminal, and Vim without config:
> 
>     $ vim -Nu NONE
> 
> Source the script (after startup):
> 
>     :source /tmp/t.vim
> 
> Finally, press:
> 
>    - `a` to enter insert mode right after the `x` character
>    - `<F3>` to complete `x`

OK, now I can reproduce it.  It happens because the screen line is
inserted later, when the popup menu is already visible.  It's easy to
skip.  Not sure if this could also happen when 'lazyredraw' is not set.

I'll just fix the problem, I'll leave it to someone else to write a
test.

>With the new example, I bisected the issue to the patch [8.1.0768](https://github.com/vim/vim/releases/tag/v8.1.0768).

That removes some redrawing.  Not really the cause of the problem.

-- 
The word "leader" is derived from the word "lead", as in the material that
bullets are made out of.  The term "leader" was popularized at about the same
time as the invention of firearms.  It grew out of the observation that the
person in charge of every organization was the person whom everyone wanted to
fill with hot lead.
   I don't recomment this; it's just a point of historical interest.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Fixed with patch Patch 8.2.5144

**Comment by lacygoill** [comment]
> Normally 'lazyredraw' should not be set, it is not normally needed.

OK, then maybe the help should mention that `'lazyredraw'` is only meant to be set temporarily.
```diff
diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index c9db88490..5cb4760f0 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -4954,6 +4954,9 @@ A jump table for the options with a short description can be found at |Q_op|.
 	executing macros, registers and other commands that have not been
 	typed.  Also, updating the window title is postponed.  To force an
 	update use |:redraw|.
+	Note that this option is only meant to be set temporarily while an
+	operation is being performed which you don't want to cause the screen
+	to be updated.
 
 			*'linebreak'* *'lbr'* *'nolinebreak'* *'nolbr'*
 'linebreak' 'lbr'	boolean	(default off)
```
---

   > That removes some redrawing.  Not really the cause of the problem.

I cannot reproduce the issue on 8.1.0767, but I can on 8.1.0768.  I've just checked again, with this downgraded code (from Vim9 to legacy):
```vim
set lazyredraw
let g:completions = ['xaaa', 'xbbb', 'xccc']
call setline(1, 'x ' . repeat('y', &columns - len('xaaa')))
call append('.', repeat([''], len(completions) + 1))
inoremap <F3> <C-R>=complete(1, g:completions) ? '' : ''<CR><C-N>
```
So, while it might not be the cause, Vim definitely started to show a spurious duplicate entry from 8.1.0768.

---

> Fixed with patch Patch 8.2.5144

Thank you.
